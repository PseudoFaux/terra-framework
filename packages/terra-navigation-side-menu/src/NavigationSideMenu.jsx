import React from 'react';
import PropTypes from 'prop-types';
import 'terra-base/lib/baseStyles';
import List from 'terra-list';
import ActionHeader from 'terra-clinical-action-header';
import ContentContainer from 'terra-content-container';
import RoutingStackDelegate from 'terra-navigation-layout/lib/RoutingStackDelegate';
import MenuItem from './_MenuItem';

const KEYCODES = {
  ENTER: 13,
};

const propTypes = {
  /**
   * An array of configuration for each menu item.
   */
  menuItems: PropTypes.arrayOf(PropTypes.shape({
    /**
     * Keys of menu items
     */
    children: PropTypes.arrayOf(PropTypes.string),
    /**
     * ID to be applied to the menu item div.
     */
    id: PropTypes.string,
    /**
     * Unique identifier that will be returned in the onChange callback when an endpoint is reached.
     */
    key: PropTypes.string.isRequired,
    /**
     * Title for the menu row and header title when selected.
     */
    title: PropTypes.string.isRequired,
  })),
  /**
   * Callback function when a menu endpoint is reached.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Delegate prop that is added by the NavigationLayout.
   */
  routingStackDelegate: RoutingStackDelegate.propType.isRequired,
  /**
   * Key of the top level menu.
   */
  initialSelectedKey: PropTypes.string.isRequired,
};

const defaultProps = {
  menuItems: [],
};

const processMenuItems = (items) => {
  const processedItems = {};
  items.forEach((item) => {
    processedItems[item.key] = { id: item.id, title: item.title, children: item.children };
  });
  return processedItems;
};

class NavigationSideMenu extends React.Component {
  constructor(props) {
    super(props);

    this.onBackClick = this.onBackClick.bind(this);
    this.onItemClick = this.onItemClick.bind(this);

    this.state = { processedItems: processMenuItems(props.menuItems), selectedKey: props.initialSelectedKey, selectedChildKey: null };
    this.previousStack = [];
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      processedItems: processMenuItems(nextProps.menuItems),
    });
  }

  onBackClick() {
    if (this.previousStack.length) {
      this.setState({ selectedKey: this.previousStack.pop(), selectedChildKey: this.state.selectedKey });
    }
  }

  onItemClick(event, key) {
    const clickedItem = this.state.processedItems[key];
    if (clickedItem.children && clickedItem.children.length) {
      this.previousStack.push(this.state.selectedKey);
      this.setState({ selectedKey: key, selectedChildKey: key });
    } else {
      this.props.onChange(event, key);
      this.setState({ selectedChildKey: key });
    }
  }

  buildListItem(key) {
    const item = this.state.processedItems[key];

    return (
      <MenuItem
        id={item.id}
        hasChevron={item.children && item.children.length > 0}
        isSelected={key === this.state.selectedChildKey}
        title={item.title}
        key={key}
        onClick={(event) => { this.onItemClick(event, key); }}
        onKeyDown={(event) => { if (event.nativeEvent.keyCode === KEYCODES.ENTER) { this.onItemClick(event, key); } }}
      />
    );
  }

  buildListContent() {
    const currentItem = this.state.processedItems[this.state.selectedKey];
    if (currentItem && currentItem.children && currentItem.children.length) {
      return <List>{currentItem.children.map(key => this.buildListItem(key))}</List>;
    }
    return null;
  }

  render() {
    const {
      menuItems,
      onChange,
      routingStackDelegate,
      initialSelectedKey,
      ...customProps
    } = this.props;
    let onBack;
    if (this.props.initialSelectedKey === this.state.selectedKey && routingStackDelegate) {
      onBack = routingStackDelegate.showParent;
    } else {
      onBack = this.onBackClick;
    }
    const currentItem = this.state.processedItems[this.state.selectedKey];
    const actionHeader = <ActionHeader onBack={onBack} title={currentItem ? currentItem.title : null} />;

    return (
      <ContentContainer {...customProps} header={actionHeader} fill>
        {this.buildListContent()}
      </ContentContainer>
    );
  }
}

NavigationSideMenu.propTypes = propTypes;
NavigationSideMenu.defaultProps = defaultProps;

export default NavigationSideMenu;
