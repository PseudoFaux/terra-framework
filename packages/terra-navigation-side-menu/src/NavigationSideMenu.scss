:local {
  .menu-list {
    box-shadow: var(--terra-navigation-side-menu-list-box-shadow, 0 7px 9px rgba(0, 0, 0, 0.2));
    padding-bottom: var(--terra-navigation-side-menu-list-padding-bottom, 0.125rem);
    padding-top: var(--terra-navigation-side-menu-list-padding-top, 0.125rem);
  }
}
