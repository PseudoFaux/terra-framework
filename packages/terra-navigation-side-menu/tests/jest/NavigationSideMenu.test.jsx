import React from 'react';

import NavigationSideMenu from '../../src/NavigationSideMenu';

describe('Layout', () => {
  it('should render a NavigationSideMenu with default props', () => {
    const result = shallow((
      <NavigationSideMenu
        onChange={() => {}}
        routingStackDelegate={{ showParent: () => {} }}
        initialSelectedKey=""
      />
    ));
    expect(result).toMatchSnapshot();
  });

  it('should render a NavigationSideMenu with selectedKey', () => {
    const result = shallow((
      <NavigationSideMenu
        menuItems={[
          { key: 'menu', title: 'Test Menu', children: ['test1', 'test2', 'test3', 'test4'] },
          { key: 'test1', title: 'Test Menu 1' },
          { key: 'test2', title: 'Test Menu 2' },
          { key: 'test3', title: 'Test Menu 3' },
          { key: 'test4', title: 'Test Menu 4' },
        ]}
        onChange={() => {}}
        routingStackDelegate={{ showParent: () => {} }}
        initialSelectedKey="menu"
      />
    ));
    expect(result).toMatchSnapshot();
  });
});
