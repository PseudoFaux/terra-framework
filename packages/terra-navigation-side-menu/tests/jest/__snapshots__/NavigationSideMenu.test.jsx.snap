// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render a NavigationSideMenu with default props 1`] = `
<ContentContainer
  fill={true}
  header={
    <ActionHeader
      className="header"
      onBack={[Function]}
      onClose={null}
      onMaximize={null}
      onMinimize={null}
      onNext={null}
      onPrevious={null}
      title={null}
    />
  }
/>
`;

exports[`Layout should render a NavigationSideMenu with selectedKey 1`] = `
<ContentContainer
  fill={true}
  header={
    <ActionHeader
      className="header"
      onBack={[Function]}
      onClose={null}
      onMaximize={null}
      onMinimize={null}
      onNext={null}
      onPrevious={null}
      title="Test Menu"
    />
  }
>
  <List
    isDivided={false}
  >
    <ListItem
      content={
        <div
          className="list-item"
          id={undefined}
        >
          Test Menu 1
        </div>
      }
      isSelected={false}
      key="test1"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <ListItem
      content={
        <div
          className="list-item"
          id={undefined}
        >
          Test Menu 2
        </div>
      }
      isSelected={false}
      key="test2"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <ListItem
      content={
        <div
          className="list-item"
          id={undefined}
        >
          Test Menu 3
        </div>
      }
      isSelected={false}
      key="test3"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <ListItem
      content={
        <div
          className="list-item"
          id={undefined}
        >
          Test Menu 4
        </div>
      }
      isSelected={false}
      key="test4"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    />
  </List>
</ContentContainer>
`;
